Add `perform_all_later` to Active Job Guide

bhumi1102 opened the pull request.

Ridhwana commented on guides/source/active_job_basics.md:
```diff
@@ -124,8 +124,68 @@ GuestsCleanupJob.perform_later(guest1, guest2, filter: 'some_filter')
 
+### Bulk Enqueuing
+
+You can Enqueue multiple jobs at once using [`perform_all_later`][]. Bulk Enqueuing reduces the number of round trips to a queue data store like Redis.
+
+`perform_all_later` is a top-level api on ActiveJob. It accepts instantiated jobs as arguments (note that this is different from `perform_later`). `perform_all_later` does call `perform` under the hood. The arguments passed to `new` will be passed on to `perform` when it's eventually called.
```
(suggestion, non-blocking): I found the order of these to be slightly confusing and took a stab at re-ordering it a bit. 

```suggestion
`perform_all_later` is a top-level api on ActiveJob that allows you to enqueue multiple jobs at once. Bulk enqueuing reduces the number of round trips to a queue data store like Redis.

`perform_all_later` accepts instantiated jobs as arguments (note that this is different from `perform_later`), and it calls  `perform` under the hood by passing the arguments from `new` to `perform`.
```

Ridhwana replied:
In addition, would be useful to say what are the benefits of enqueuing multiple jobs or why you'd want to do that?

Ridhwana commented on guides/source/active_job_basics.md:
```diff
@@ -124,8 +124,68 @@ GuestsCleanupJob.perform_later(guest1, guest2, filter: 'some_filter')
+
+`perform_all_later` logs the number of jobs successfully enqueued, something like `Enqueued 3 jobs to Async (3 GuestsCleanupJob)` if it was called with `ActiveJob.perform_all_later(job1, job2, job3)`. 
+
+The return value of `perform_all_later` is `nil`. Note that this is different from `perform_later`. This may be enhanced to return the number of successfully enqueued job in the future, but for now it's `nil`.
```
(suggestion, non-blocking): Not sure that the last part of the sentence is needed.

```suggestion
The return value of `perform_all_later` is `nil`. Note that this is different from `perform_later`. This may be enhanced to return the number of successfully enqueued job in the future.
```

Ridhwana commented on guides/source/active_job_basics.md:
````diff
@@ -124,8 +124,68 @@ GuestsCleanupJob.perform_later(guest1, guest2, filter: 'some_filter')
+```
+#### Callbacks
+
+Note that `perform_all_later` does not trigger callbacks on the individual jobs. This is inline with other ActiveRecord bulk methods. Since callbacks run on individual jobs, they can't take advantage of the bulk nature of this method. Also there isn't a meaningful semantic for something like `around_enqueue` callback.
````
(non-blocking, suggestion): The series of short sentences threw me off a bit but maybe this could be personal preference, I'm not sure. Feel free to ignore it if you do not agree. 

```suggestion
`perform_all_later` does not trigger callbacks on the individual jobs (this behavior is in line with other ActiveRecord bulk methods). Callbacks run on individual jobs, hence they can't take advantage of the bulk nature of this method, and there isn't a meaningful semantic for something like `around_enqueue` callback.
```

I'm not sure that I fully understand this portion of the sentence `and there is not a meaningful semantic for something like `around_enqueue` callback.`

Ridhwana commented on guides/source/active_job_basics.md:
```diff
@@ -124,8 +124,68 @@ GuestsCleanupJob.perform_later(guest1, guest2, filter: 'some_filter')
+
+#### Queue Backend Support
+
+For `perform_all_later`, bulk enqueuing needs to be backed by the queue backend.
```
Maybe here we can link to the section in the guides on Backends?

```suggestion
For `perform_all_later`, bulk enqueuing needs to be backed by the [queue backend](https://edgeguides.rubyonrails.org/active_job_basics.html#backends).
```

Ridhwana commented on guides/source/active_job_basics.md:
```diff
@@ -406,6 +466,30 @@ end
 [`around_perform`]: https://api.rubyonrails.org/classes/ActiveJob/Callbacks/ClassMethods.html#method-i-around_perform
 [`after_perform`]: https://api.rubyonrails.org/classes/ActiveJob/Callbacks/ClassMethods.html#method-i-after_perform
 
+### Callbacks with Bulk Enqueuing
```
We have the "Callbacks" section above and the "Callbacks with Bulk Enqueuing" section here. Should they be merged into something like "Callbacks and Events" or perhaps we can keep them as two separate sections like "Callbacks" and "Events"?

Ridhwana commented on guides/source/active_job_basics.md:
```diff
@@ -406,6 +466,30 @@ end
+
+ However, the `perform_all_later` method does fire an [`enqueue_all.active_job`][] event which you can subscribe to using `ActiveSupport::Notifications`. 
+
+The event payload contains the adapter name, enqueued jobs count, the list of job instances, etc. `successfully_enqueued?` can be used to find out if a given job was successfully enqueued. 
```
(non-blocking, suggestion): I didn't realize that the sentence ended with etc. and sho I tried to read both these sentences as one. It felt like it impacted the readability because the new sentence did not start with a capital letter. 

```suggestion
The event payload contains the adapter name, enqueued jobs count, the list of job instances, etc. A method `successfully_enqueued?` can be used to find out if a given job was successfully enqueued. 
```

Ridhwana replied:
Maybe we can also provide a link to the `successfully_enqueued?` method?

bhumi1102 replied:
Sure. I've linked it though that api is pretty bare https://api.rubyonrails.org/classes/ActiveJob/Core.html#method-i-successfully_enqueued-3F

Ridhwana reviewed (commented):
This looks great! ✨ I left a few suggestions and comments and marked them as optional etc.

bhumi1102 commented:
I've addressed all review feedback. Anything else before moving to PR on rail/rails? Oh, I'm up for moving the location of section 3.3, let me know if you have a strong preference to where?

carlosantoniodasilva commented on guides/source/active_job_basics.md:
```diff
@@ -406,6 +465,14 @@ end
+
+When enqueueing jobs in bulk using `perform_all_later`, callbacks such as `around_enqueue`, will not be triggered on the individual jobs.
+
+ However, the `perform_all_later` method does fire an [`enqueue_all.active_job`](active_support_instrumentation.html#enqueue-all-active-job) event which you can subscribe to using `ActiveSupport::Notifications`.
```
Nit: extra space.

```suggestion
However, the `perform_all_later` method does fire an [`enqueue_all.active_job`](active_support_instrumentation.html#enqueue-all-active-job) event which you can subscribe to using `ActiveSupport::Notifications`.
```

carlosantoniodasilva reviewed (commented):
Looking great! I did another quick read through, and had another few minor suggestions, one typo in the multiple classes example.

> I'm up for moving the location of section 3.3, let me know if you have a strong preference to where?

Not necessarily a strong preference, but there's a couple of reasons we might want to consider moving it to at least after the callbacks section:

- It has its own little section on callbacks, which links down the guide, and then there's a sub-section on callbacks with bulk enqueuing as well. Moving it down probably means we can merge the two into a single one.
- It explains the need to be backed by a backend queue, but we haven't gotten to the point of explaining what backend queues are yet. Moving it down would put it after that concept explanation.

I can also see why keep it early in the guide, within the section we explain about how to create and enqueue jobs. But it could be treated more as an "advanced" concept / form of enqueueing as well.

That doesn't need to stop you from moving the PR to rails/rails, maybe could even ask the broader community for feedback on that specifically if still unsure. We might still review the Active Job guide as a whole later on as well.

bhumi1102 commented on guides/source/active_job_basics.md:
```diff
@@ -127,6 +127,65 @@ That's it!
+notify_job = NotifyGuestsJob.new(guest)
+
+# Enqueues job instances from multiple classes at once
+ApplicationJob.perform_all_later(cleanup_job, export_job, notify_job)
```
Thanks for catching that!

carlosantoniodasilva commented on guides/source/active_job_basics.md:
```diff
@@ -406,6 +410,72 @@ end
 [`around_perform`]: https://api.rubyonrails.org/classes/ActiveJob/Callbacks/ClassMethods.html#method-i-around_perform
 [`after_perform`]: https://api.rubyonrails.org/classes/ActiveJob/Callbacks/ClassMethods.html#method-i-after_perform
 
+Please note that When enqueueing jobs in bulk using `perform_all_later`, callbacks such as `around_enqueue`, will not be triggered on the individual jobs. See [Bulk Enqueuing Callbacks](#bulk-enqueue-callbacks).
```
```suggestion
Please note that when enqueueing jobs in bulk using `perform_all_later`, callbacks such as `around_enqueue`, will not be triggered on the individual jobs. See [Bulk Enqueuing Callbacks](#bulk-enqueue-callbacks).
```

bhumi1102 commented:
> I can also see why keep it early in the guide, within the section we explain about how to create and enqueue jobs. But it could be treated more as an "advanced" concept / form of enqueueing as well.
> 
> That doesn't need to stop you from moving the PR to rails/rails, maybe could even ask the broader community for feedback on that specifically if still unsure. We might still review the Active Job guide as a whole later on as well.

I moved the whole section 3.3 to a new top level section 8. And linked the new section within 3.3. Also cleaned up the 'callbacks' docs in multple places. Seems good for now. Going over the entire Active Job guide later on sounds good!

Will create a rails/rails PR now!
